Feat: Allow saving popup contents to temp_registers when quitting

jakubbortlik opened the pull request:
This MR addresses #177. The user can now specify a list of registers in `settings.popup.temp_registers` to which the contents of editable popups will be saved when exiting the popup.

A couple of notes on things that might be done differently:
1. If no `temp_registers` are defined, nothing happens. Instead, when quitting a popup without a temp_register being set, the user could be notified about the possibility as mentioned in the original feature request #177.
2. The `temp_registers` are filled even when triggering `settings.popup.perform_action` because this also triggers the `exit` function with the callback `lua/gitlab/state.lua:259` that does the popup saving. If this behaviour is undesirable, the `popup_opts` could be extended to make it possible not to save upon `perform_action`. However, I find this to be a nice side effect to the main purpose of `temp_registers`, as it makes it possible to get rid of the `backup_register` hotfix.

jakubbortlik commented:
Hi @harrisoncramer, what about instead of completely removing the `<ESC>` keybinding we'd just replace it with something that is typically used in plugins to close popup windows, like `gq`? I'm using just `q` and I'm happy with it as I rarely need to record to a register while writing a comment.

With the keymap removed, will the saving to `temp_register` even take place? I'd think we should keep the keymap (and change it from `<ESC>` to something else) even if we'd implement saving to temp files.

jakubbortlik commented:
By the way, @petRUShka, I'm using `<ZZ>` as my `perform_action` keymap which effectively makes it behave like saving the commit message you mentioned: I "save" the buffer and the comment is applied.

harrisoncramer commented:
Hey both,

Thanks for the feedback. Believe ultimately that @petRUShka is right, and that the plugin shouldn't be creating new abstractions for things like closing buffers or tabs that are already part of Neovim. 

Instead of having a keybinding for this, users should now just close the popups like they would any other buffer. To address your point @jakubbortlik, I've added an autocommand to the `state.lua` file that will trigger whenever a popup unloads. If you want to add custom saving functionality like in this plugin, this would be the place to do it, as this autocommand is called already when the buffer unloads:

```lua
  vim.api.nvim_create_autocmd("BufUnload", {
    buffer = popup.bufnr,
    callback = function()
      local text = u.get_buffer_text(popup.bufnr)
      -- Do something with the text here...
      exit(popup, opts)
    end,
  })
```

jakubbortlik commented:
> Hey both, thanks for the feedback.
> 
> The plugin shouldn't be creating new abstractions for things like closing buffers or tabs that are already part of Neovim, so I've removed the ability to add one for popups.
> 
> Instead of having a keybinding for this, users should now just close the popups like they would any other buffer. To address your point @jakubbortlik, I've added an autocommand to the `state.lua` file that will trigger whenever a popup unloads. If you want to add custom saving functionality, this would be the place to do it, as this autocommand is called already when the buffer unloads:
> 
> ```lua
>   vim.api.nvim_create_autocmd("BufUnload", {
>     buffer = popup.bufnr,
>     callback = function()
>       local text = u.get_buffer_text(popup.bufnr)
>       -- Do something with the text here...
>       exit(popup, opts)
>     end,
>   })
> ```
> 
> I still support this MR and think we should add it to the plugin!
> 
> To the other points about saving all comments as drafts, or saving in real-time as the user inputs text, I don't anticipate adding this to the plugin at any point soon, this would actually be very complex to implement and I'm not sure how valuable it'll be. Supporting state for parallel workstreams in different comments/buffers is not a can of worms I'm keen to open up.
> 
> @jakubbortlik Please let me know when this is ready for review and I'll take a look but we should try to get #230 merged first.

Hi @harrisoncramer, thanks for your input. As you say, I'm first focusing on finishing #230 and I'll get back to this after that.

jakubbortlik commented:
Hi @harrisoncramer, I think the feature works as expected now, although I'd like to suggest one small change to the `autocommand`. If we changed `BufUnload` to `BufLeave`, then a user (me :smile:) would be able to create this mapping to quit a popup without saving the contents to the `temp_registers`:
```lua
vim.keymap.set("n", "ZQ", function()
  local reg_backup = gitlab.state.settings.popup.temp_registers
  gitlab.state.settings.popup.temp_registers = {}
  vim.cmd("quit")
  gitlab.state.settings.popup.temp_registers = reg_backup
end)
```
This is inspired by the behaviour of normal `ZQ` in that it quits "without checking for changes". For some reason, this mapping does not work when `BufUnload` is used. In that case, `ZQ` still saves to the registers.
